sympy/core/tests/test_relational.py

```diff
@@ -1026,18 +1028,32 @@ def test_rel_args():
                 raises(TypeError, lambda: Relational(b, v, op))
 
 
-def test_Equality_rewrite_as_Add():
+def test_Equality_lhs_rhs():
+    # XXX we can be strict and make every routine that want to treat
+    # Eq like Eqn for purpose of solving explicitly test for Eq
+    #     if isinstance(eq, Eq):
+    #        eq = eq.lhs_rhs()
+    # or allow Expr to pass
+    #     eq = eq.lhs_rhs() # valid for Eq or Expr
```

Comment:
I think it is better to be strict. Code that does not know whether `eq` is an `Eq` or an `Expr` is always fragile. In any situation where either `Eq` or `Expr` is allowed then a conversion should be done at the earliest opportunity either like:
```python
if isinstance(eq, Eq):
     eq = eq.lhs_rhs()
elif isinstance(eq, Expr):
     eq = eq
else:
    raise Error
# Now we know we have Expr
```
Or it should be like:
```python
if isinstance(eq, Eq):
    lhs, rhs = eq.lhs, eq.rhs
elif isinstance(eq, Expr):
    lhs, rhs = eq, S.Zero
else:
    raise Error
# Now we know we have a pair of Expr
```
Or
```python
if isinstance(eq, Eq):
    eq = eq
elif isinstance(eq, Expr):
    eq = Eq(eq, 0)
else:
    raise Error
# Now we know we have Eq
```
I don't think that burying this logic into a convenience method helps for clarity of code.

In general we should minimise having situations where either `Eq` or `Expr` can be used or at least try to localise very carefully where the conversions are done if needed.